File: app/src/main/java/education/mahmoud/quranyapp/feature/show_tafseer/TafseerDetails.java
package education.mahmoud.quranyapp.feature.show_tafseer;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Spinner;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import org.koin.java.KoinJavaComponent;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;
import education.mahmoud.quranyapp.R;
import education.mahmoud.quranyapp.data_layer.Repository;
import education.mahmoud.quranyapp.data_layer.local.room.AyahItem;
import education.mahmoud.quranyapp.data_layer.local.room.SuraItem;
import education.mahmoud.quranyapp.feature.download.DownloadActivity;
import education.mahmoud.quranyapp.utils.Data;

public class TafseerDetails extends Fragment {

    private static final String TAG = "TafseerDetails";

    @BindView(R.id.spSuraTafser)
    Spinner spSuraTafser;
    @BindView(R.id.spAyahTafser)
    Spinner spAyahTafser;
    @BindView(R.id.rvTafeer)
    RecyclerView rvTafeer;
    @BindView(R.id.tvNoDataInTafseer)
    TextView tvNoDataInTafseer;

    Unbinder unbinder;
    private Repository repository = KoinJavaComponent.get(Repository.class);
    private SuraItem sura;
    private List<AyahItem> suraAyahsTafseer;
    private TafseerAdapter adapter;

    public static TafseerDetails newInstance(String title) {
        TafseerDetails frag = new TafseerDetails();
        return frag;
    }

    /**
     * used onCreateView when providing a custom UI for fragment
     *
     * @param inflater
     * @param container
     * @param savedInstanceState
     * @return
     */
    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_tafseer_details, container, false);
        unbinder = ButterKnife.bind(this, view);

        fillSpinners();
        initRv();
        return view;
    }

    private void initRv() {
        LinearLayoutManager manager = new LinearLayoutManager(getContext());
        rvTafeer.setLayoutManager(manager);
        adapter = new TafseerAdapter();
        rvTafeer.setHasFixedSize(true);
        rvTafeer.setAdapter(adapter);
        Log.d(TAG, "initRv: ");
    }

    private void fillSpinners() {

        List<String> suraNames = Arrays.asList(Data.SURA_NAMES);
        ArrayAdapter<String> startAdapter = new ArrayAdapter<>(getContext(), android.R.layout.simple_dropdown_item_1line, suraNames);
        spSuraTafser.setAdapter(startAdapter);
        spSuraTafser.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
                String suraName = (String) spSuraTafser.getSelectedItem();
                suraAyahsTafseer = repository.getAllAyahOfSurahIndexForTafseer(l + 1);
                Log.d(TAG, "onItemSelected: " + suraAyahsTafseer.size() + " & " + l);
                // update adapter
                if (suraAyahsTafseer.size() > 0) {
                    foundState();
                    adapter.setTafseerList(suraAyahsTafseer);
                   // Log.d(TAG, "onItemSelected: " + suraName);

                    List<String> ayahsNums = creatAyahsNumList(suraAyahsTafseer.size()); // size() ->  n of ayahs
                    ArrayAdapter<String> adapter = new ArrayAdapter<>(getContext() ,android.R.layout.simple_dropdown_item_1line,ayahsNums );
                    spAyahTafser.setAdapter(adapter);
                    spAyahTafser.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
                        @Override
                        public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
                            rvTafeer.scrollToPosition((int) l);
                            Log.d(TAG, "onItemSelected: TAFSEEER POS " + l);
                        }

                        @Override
                        public void onNothingSelected(AdapterView<?> adapterView) {

                        }
                    });

                } else {
                    notFoundState();
                }
            }

            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {

            }
        });
    }

    private List<String> creatAyahsNumList(int i) {
        List<String> ayahsNums = new ArrayList<>();
        for (int j = 1; j <= i  ; j++) {
            ayahsNums.add(String.valueOf(j));
        }
        return ayahsNums;
    }

    private void foundState() {
        rvTafeer.setVisibility(View.VISIBLE);
        spSuraTafser.setVisibility(View.VISIBLE);
        spAyahTafser.setVisibility(View.VISIBLE);
        tvNoDataInTafseer.setVisibility(View.GONE);
    }

    private void notFoundState() {
        rvTafeer.setVisibility(View.GONE);
        spSuraTafser.setVisibility(View.INVISIBLE);
        spAyahTafser.setVisibility(View.INVISIBLE);
        tvNoDataInTafseer.setVisibility(View.VISIBLE);
    }



    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    @OnClick(R.id.tvNoDataInTafseer)
    public void onViewClicked() {
        Intent openAcivity = new Intent(getContext(), DownloadActivity.class);
        getContext().startActivity(openAcivity);

    }
}

File: app/src/main/java/education/mahmoud/quranyapp/feature/show_tafseer/TafseerDetails.kt
package education.mahmoud.quranyapp.feature.show_tafseer

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.AdapterView.OnItemSelectedListener
import android.widget.ArrayAdapter
import com.zhuinden.fragmentviewbindingdelegatekt.viewBinding
import education.mahmoud.quranyapp.R
import education.mahmoud.quranyapp.base.BaseFragment
import education.mahmoud.quranyapp.databinding.FragmentTafseerDetailsBinding
import education.mahmoud.quranyapp.datalayer.QuranRepository
import org.koin.android.ext.android.inject

class TafseerDetails : BaseFragment() {
    private val quranRepository: QuranRepository by inject()

    private val adapter = TafseerAdapter()

    private val binding by viewBinding(FragmentTafseerDetailsBinding::bind)

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View? {
        return inflater.inflate(R.layout.fragment_tafseer_details, container, false)
    }

    override fun initViews(view: View) {
        super.initViews(view)
        fillSpinners()
        initRv()
    }

    override fun setClickListeners() {
        super.setClickListeners()
        binding.tvNoDataInTafseer.setOnClickListener {}
    }

    private fun initRv() {
        with(binding) {
            rvTafeer.setHasFixedSize(true)
            rvTafeer.adapter = (adapter)
        }
    }

    private fun fillSpinners() {
        val suraNames = quranRepository.surasNames
        val startAdapter =
            ArrayAdapter(requireContext(), android.R.layout.simple_dropdown_item_1line, suraNames)
        with(binding) {
            spSuraTafser.adapter = startAdapter
            try {
                spSuraTafser.onItemSelectedListener = object : OnItemSelectedListener {
                    override fun onItemSelected(
                        adapterView: AdapterView<*>?,
                        view: View?,
                        i: Int,
                        l: Long,
                    ) {
                        val ayahs = quranRepository.getAllAyahOfSurahIndexForTafseer(l + 1)
                        Log.d(TAG, "onItemSelected: " + ayahs.size + " & " + l)
                        // update adapter
                        if (ayahs.isNotEmpty()) {
                            foundState()
                            adapter.setTafseerList(ayahs)
                            // Log.d(TAG, "onItemSelected: " + suraName);
                            val ayahsNums = creatAyahsNumList(ayahs.size) // size() ->  n of ayahs
                            val adapter = ArrayAdapter(
                                requireContext(),
                                android.R.layout.simple_dropdown_item_1line,
                                ayahsNums
                            )
                            spAyahTafser.adapter = adapter
                            spAyahTafser.onItemSelectedListener = object : OnItemSelectedListener {
                                override fun onItemSelected(
                                    adapterView: AdapterView<*>?,
                                    view: View?,
                                    i: Int,
                                    l: Long,
                                ) {
                                    rvTafeer.scrollToPosition(l.toInt())
                                    Log.d(TAG, "onItemSelected: TAFSEEER POS $l")
                                }

                                override fun onNothingSelected(adapterView: AdapterView<*>?) {}
                            }
                        } else {
                            notFoundState()
                        }
                    }

                    override fun onNothingSelected(adapterView: AdapterView<*>?) {}
                }
            } catch (e: Exception) {
            }
        }
    }

    private fun creatAyahsNumList(i: Int): List<String> {
        val ayahsNums: MutableList<String> = ArrayList()
        for (j in 1..i) {
            ayahsNums.add(j.toString())
        }
        return ayahsNums
    }

    private fun foundState() {
        with(binding) {
            rvTafeer.visibility = View.VISIBLE
            spSuraTafser.visibility = View.VISIBLE
            spAyahTafser.visibility = View.VISIBLE
            tvNoDataInTafseer.visibility = View.GONE
        }
    }

    private fun notFoundState() {
        with(binding) {
            rvTafeer.visibility = View.GONE
            spSuraTafser.visibility = View.INVISIBLE
            spAyahTafser.visibility = View.INVISIBLE
            tvNoDataInTafseer.visibility = View.VISIBLE
        }
    }

    companion object {
        private const val TAG = "TafseerDetails"
        fun newInstance(title: String?): TafseerDetails {
            return TafseerDetails()
        }
    }
}
